Readme.md
# Random Color

[![Coverage Status](https://coveralls.io/repos/github/sterlingwes/RandomColor/badge.svg?branch=master)](https://coveralls.io/github/sterlingwes/RandomColor?branch=master)

Random color is a simple javascript module for generating random colors that contrast enough to distinguish between one another, and are not too loud (by default, all colours can have readable black/dark font overlayed).

[Example](https://cdn.rawgit.com/sterlingwes/RandomColor/5a65e8ff/example/index.html)

Useful, for example, in charts and diagrams where X number of series need to be distinguished by color.

[See changelog](/CHANGELOG.md).

## Getting it

### Node.js
`npm install rcolor`

### Browser
Clone this repo / copy `dist/rcolor.min.js`

*Note* that the `dist/rcolor.js` file is a CommonJS export (the node.js build)

## Using it

With the browser build:

```javascript
window.rcolor()
// > "#79f2ab"
```

With the node build:

```javascript
const rcolor = require('rcolor')
rcolor() // "#79f2ab"
```

This module holds the initial random number seed as internal state. To regenerate the random number seed, call the `reSeed` method: `rcolor.reSeed()`.

## Options / Configuration

You can specify the `hue`, `saturation` and `value`. The following are the defaults:

```javascript
const goldenRatio = 0.618 // ... truncated

rcolor({
  hue: (Math.random() + goldenRatio) % 1,
  saturation: 0.5,
  value: 0.95
})
```

## Pre-v1.0

If you used this module prior to NPM availability (between 2013 and 2017), you can find that source (including the Dart version) on the [legacy branch](https://github.com/sterlingwes/RandomColor/tree/legacy).

## License

Free to use and distribute under the MIT license.

## Sorry, Colour

Apologies in advance to my commonwealth friends, if you find 'color' vs. 'colour'. It was a hard decision for a Canadian. :)

Wes Johnson
[@SterlingWes](http://twitter.com/SterlingWes)

CHANGELOG.md
## v2.0.0

Fundamentally changed how randomly generated colors are seeded which fixes [the issue of colors of close proximity](https://github.com/sterlingwes/RandomColor/issues/5). Thank you [@aparshin](https://github.com/aparshin)!

This is a breaking change given how the generated output will differ. It also does not augment the `hue` if one is passed as a property on the input argument.


src/index.test.js
/* global describe, it */
const expect = require('expect.js');
const rcolor = require('./index');

const colors = [
  // hsv, rgb
  ['black', [0, 0, 0], [0, 0, 0]],
  ['white', [0, 0, 1], [255, 255, 255]],
  ['red', [0, 1, 1], [255, 0, 0]],
  ['lime', [deg(120), 1, 1], [0, 255, 0]],
  ['blue', [deg(240), 1, 1], [0, 0, 255]],
  ['yellow', [deg(60), 1, 1], [255, 255, 0]],
  ['cyan', [deg(180), 1, 1], [0, 255, 255]],
  ['magenta', [deg(300), 1, 1], [255, 0, 255]],
  ['silver', [0, 0, 0.75], [191, 191, 191]], // 192?
  ['gray', [0, 0, 0.5], [127, 127, 127]], // 128?
  ['maroon', [0, 1, 0.5], [127, 0, 0]], // 128?
  ['olive', [deg(60), 1, 0.5], [127, 127, 0]], // 128?
  ['green', [deg(120), 1, 0.5], [0, 127, 0]], // 128?
  ['purple', [deg(300), 1, 0.5], [127, 0, 127]], // 128
  ['teal', [deg(180), 1, 0.5], [0, 127, 127]],
  ['navy', [deg(240), 1, 0.5], [0, 0, 127]],
];

describe('rcolor', function() {
  it('should return hex colors by default', function() {
    const hex = rcolor();
    expect(/^#[a-z0-9]{6}/.test(hex)).to.be(true);
  });

  describe('hsvToRgb', () => {
    const fromHsv = require('./hsvToRgb');

    colors.forEach(([colorName, hsv, expectedRgb]) => {
      it(`should be ${colorName}`, () => {
        const rgb = fromHsv.apply(null, hsv);
        expect(rgb).to.eql(expectedRgb);
      });
    });
  });

  describe('padHex', () => {
    it('should not pad if greater than 2', () => {
      const padHex = require('./padHex');
      expect(padHex('000')).to.be('000');
    });
  });
});

function deg(degrees) {
  return degrees / 360;
}


dart/rcolor.dart
// Free to use & distribute under the MIT license
// Michael Moreno
//
// Dart implementation of Wes Johnson's RandomColor JavaScript module 
// inspired by http://martin.ankerl.com/2009/12/09/how-to-create-random-colors-programmatically/

library rcolor;

import "dart:math";

class RColor {
  final double _goldenRatio = 0.618033988749895;
  
  int _hexWidth;
  double _hue;
  
  RColor() {
    _hexWidth = 2;
    
    _hue = new Random().nextDouble();
  }
  
  List _hsvToRgb(double h, double s, double v){
    int h_i = (h * 6).floor();
    double f = h * 6 - h_i;
    double p = v * (1 - s);
    double q = v * (1 - f * s);
    double t = v * (1 - (1 - f) * s);
    
    double r,g,b = 255.0;
    
    switch(h_i) {
      case 0:
        r = v; g = t; b = p; break;
      case 1:
        r = q; g = v; b = p; break;
      case 2:
        r = p; g = v; b = t; break;
      case 3:
        r = p; g = q; b = v;  break;
      case 4:
        r = t; g = p; b = v;  break;
      case 5:
        r = v; g = p; b = q; break;
    }
    
    return [(r * 256).floor(), (g * 256).floor(), (b * 256).floor()];
  }
  
  String getHex([double saturation = 0.5, double value = 0.95]) {
    _updateHue();
    
    List rgb = _hsvToRgb(_hue, saturation, value);
    
    return "#" + rgb[0].toRadixString(16) + 
                 rgb[1].toRadixString(16) + 
                 rgb[2].toRadixString(16);
  }
  
  List getRGB([double saturation = 0.5, double value = 0.95]) {
    _updateHue();
    
    return _hsvToRgb(_hue, saturation, value);
  }
  
  void _updateHue() {
    _hue+= _goldenRatio;
    _hue%= 1;
  }
}

void main(){}

rcolor.js
// Free to use & distribute under the MIT license
// Wes Johnson (@SterlingWes)
//
// inspired by http://martin.ankerl.com/2009/12/09/how-to-create-random-colors-programmatically/

(function (root, factory) {
    if (typeof exports === 'object') {
        module.exports = factory;
    } else if (typeof define === 'function' && define.amd) {
        define(factory);
    } else {
        root.RColor = factory();
    }
}(this, function () {

	var RColor = function() {
		this.hue			= Math.random(),
		this.goldenRatio 	= 0.618033988749895;
		this.hexwidth		= 2;
	};

	RColor.prototype.hsvToRgb = function (h,s,v) {
		var	h_i	= Math.floor(h*6),
			f 	= h*6 - h_i,
			p	= v * (1-s),
			q	= v * (1-f*s),
			t	= v * (1-(1-f)*s),
			r	= 255,
			g	= 255,
			b	= 255;
		switch(h_i) {
			case 0:	r = v, g = t, b = p;	break;
			case 1:	r = q, g = v, b = p;	break;
			case 2:	r = p, g = v, b = t;	break;
			case 3:	r = p, g = q, b = v;	break;
			case 4: r = t, g = p, b = v;	break;
			case 5: r = v, g = p, b = q;	break;
		}
		return [Math.floor(r*256),Math.floor(g*256),Math.floor(b*256)];
	};
	
	RColor.prototype.padHex = function(str) {
		if(str.length > this.hexwidth) return str;
		return new Array(this.hexwidth - str.length + 1).join('0') + str;
	};
	
	RColor.prototype.get = function(hex,saturation,value) {
		this.hue += this.goldenRatio;
		this.hue %= 1;
		if(typeof saturation !== "number")	saturation = 0.5;
		if(typeof value !== "number")		value = 0.95;
		var rgb = this.hsvToRgb(this.hue,saturation,value);
		if(hex)
			return "#" +  this.padHex(rgb[0].toString(16))
						+ this.padHex(rgb[1].toString(16))
						+ this.padHex(rgb[2].toString(16));
		else 
			return rgb;
	};

	return RColor;

}));